From 39f9caba2a4c7698b094976015465e8783a10b10 Mon Sep 17 00:00:00 2001
From: Arulselvan M <arulselvan.m@intel.com>
Date: Thu, 16 Jul 2015 05:12:59 +0530
Subject: [PATCH] ASoC: SST: [Workaround] To serve missed LPE interrupt

Due to hardware bug, though the done and busy bit are set
but interrupt is not triggered.This patch provides workaround to serve
the missed interrupt.

Change-Id: I4232b6fb8d14d96d8c411ba2d6854c51503c1560
Tracked-On: https://jira01.devtools.intel.com/browse/GMINL-14759
Signed-off-by: Arulselvan M <arulselvan.m@intel.com>
---
 sound/soc/intel/sst/sst.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/sound/soc/intel/sst/sst.c b/sound/soc/intel/sst/sst.c
index abc0b6b614ea..2b4340b6b341 100644
--- a/sound/soc/intel/sst/sst.c
+++ b/sound/soc/intel/sst/sst.c
@@ -132,9 +132,10 @@ static irqreturn_t intel_sst_interrupt_mrfld(int irq, void *context)
 	irqreturn_t retval = IRQ_HANDLED;
 	unsigned long irq_flags;
 
+	header.full = sst_shim_read64(drv->shim, drv->ipc_reg.ipcx);
 	/* Interrupt arrived, check src */
 	isr.full = sst_shim_read64(drv->shim, SST_ISRX);
-	if (isr.part.done_interrupt) {
+	if (isr.part.done_interrupt || header.p.header_high.part.done) {
 		/* Clear done bit */
 		spin_lock_irqsave(&drv->ipc_spin_lock, irq_flags);
 		header.full = sst_shim_read64(drv->shim,
@@ -151,7 +152,8 @@ static irqreturn_t intel_sst_interrupt_mrfld(int irq, void *context)
 		queue_work(drv->post_msg_wq, &drv->ipc_post_msg.wq);
 		retval = IRQ_HANDLED;
 	}
-	if (isr.part.busy_interrupt) {
+	header.full = sst_shim_read64(drv->shim, drv->ipc_reg.ipcd);
+	if (isr.part.busy_interrupt || header.p.header_high.part.busy) {
 		spin_lock_irqsave(&drv->ipc_spin_lock, irq_flags);
 		imr.full = sst_shim_read64(drv->shim, SST_IMRX);
 		imr.part.busy_interrupt = 1;
-- 
1.9.1

